com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/ThreadsRequestHandler.java

```diff
@@ -168,4 +193,140 @@ private void checkThreadRunningAndRecycleIds(ThreadReference thread, IDebugAdapt
+                // do another step of the same kind.
+                if (locationShouldBeFiltered(thread, context) || shouldDoExtraStepInto(thread)) {
+                    DebugUtility.stepInto(thread, debugSession.getEventHub(), context.getDebugFilters().stepFilters);
+                    return;
+                }
+            }
+        }
+        context.sendEvent(new Events.StoppedEvent("step", thread.uniqueID()));
+    }
+
+    /**
+     * Return true if the StepEvent's location is a Method that the user has indicated to filter.
+     */
+    private boolean locationShouldBeFiltered(ThreadReference thread, IDebugAdapterContext context) {
+        Location originalLocation = getOriginalStepLocation(thread);
+        Location currentLocation = null;
+        StackFrame topFrame = DebugUtility.getTopFrame(thread);
+        if (topFrame != null) {
+            currentLocation = topFrame.location();
+        }
+        if (originalLocation == null || currentLocation == null) {
+            return false;
+        }
+        return !methodShouldBeFiltered(originalLocation.method(), context) && methodShouldBeFiltered(currentLocation.method(), context);
+    }
+
+    private boolean methodShouldBeFiltered(Method method, IDebugAdapterContext context) {
+        if ((context.getDebugFilters().skipStaticInitializers && method.isStaticInitializer())
+                || (context.getDebugFilters().skipSynthetics && method.isSynthetic())
+                || (context.getDebugFilters().skipConstructors && method.isConstructor())) {
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Check if the current top stack is same as the original top stack.
+     */
+    private boolean shouldDoExtraStepInto(ThreadReference thread) {
+        if (getOriginalStackDepth(thread) != DebugUtility.getFrameCount(thread)) {
+            return false;
+        }
+        Location originalLocation = getOriginalStepLocation(thread);
+        if (originalLocation == null) {
+            return false;
+        }
+        Location currentLocation = DebugUtility.getTopFrame(thread).location();
+        Method originalMethod = originalLocation.method();
+        Method currentMethod = currentLocation.method();
+        if (!originalMethod.equals(currentMethod)) {
+            return false;
+        }
+        if (originalLocation.lineNumber() != currentLocation.lineNumber()) {
+            return false;
+        }
+        return true;
+    }
+
+    private ThreadState getThreadState(ThreadReference thread) {
+        if (thread  == null) {
+            return null;
+        }
+        long threadId = thread.uniqueID();
+        if (!threadStates.containsKey(threadId)) {
+            threadStates.put(threadId, new ThreadState());
+        }
+        return threadStates.get(threadId);
+    }
+
+    private void setPendingStepKind(ThreadReference thread, int kind) {
```

Comment:
Should be another ThreadUtility class if you already put something into the DebugUtility.

Reply:
At the ThreadsRequestHandler constructor, i create a thread state map `private Map<Long, ThreadState> threadStates = new HashMap<>();`. This method is used to update the state map.